File: services/app/services/llms/gemini.py
import json
from google import genai
from google.genai import types
from io import BytesIO
from PIL import Image
import boto3
import uuid
import time

from app.config import settings
from app.services.llms.base import Base

NAME = "GeminiServices"
MINUTE_SECONDS = 60
DAY_SECONDS = 86400

class GeminiServices(Base):
    def __init__(self):
        super().__init__(NAME)

        self.name = NAME
        self.rpm =10 
        self.rpd =250
        self.tpm = 250000
        
        self.API_KEY = settings.GOOGLE_GEMINI_APIKEY
        self.API_KEYS = [settings.GOOGLE_GEMINI_APIKEY]
        self.MODEL = "gemini-2.5-flash"
        self.api_key_records = []

        self.AWS_ACCESS_KEY = settings.AWS_ACCESS_KEY
        self.AWS_SECRET_ACCESS_KEY = settings.AWS_SECRET_ACCESS_KEY
        self.AWS_REGION = settings.AWS_REGION
        self.AWS_BUCKET = settings.AWS_BUCKET

    def get_api_key(self):
        
        api_key = self.API_KEY

        for key in self.API_KEYS:

            request_key = f"{self.name}_{key}_request_key"

            data = list(json.loads(self.redis_cli.get(request_key) or '[]'))

            self.api_key_records = data

            if not data:
                api_key = key

            today_total_requests = 0
            current_min_requests = 0
            current_min_tokens = 0

            for d in data:
                if d['time'] >= int(time.time()) - MINUTE_SECONDS:
                    current_min_requests += d['count']
                    current_min_tokens += d['tokens']
                if d['time'] >= int(time.time()) - DAY_SECONDS:
                    today_total_requests += d['count']

            if today_total_requests == 0 and current_min_requests == 0 and current_min_tokens == 0:
                api_key = key

            if today_total_requests > self.rpd or current_min_requests > self.rpm or current_min_tokens > self.tpm:
                continue

            api_key = key

        return api_key
     
    def _generate(self, request):
        
        try: 


            prompt = request.get('prompt', "")
            format = request.get('format') or "text"

            response_format = None
            if format == "json":
                response_format = "application/json"
            elif format == "text":
                response_format = "text/plain"

            API_KEY = self.get_api_key()

            client = genai.Client(api_key=API_KEY)

            response = client.models.generate_content(
                model=self.MODEL,
                contents=prompt,
                config=types.GenerateContentConfig(
                    candidate_count=1,
                    response_mime_type=response_format
                ),
            )
            
            if not response.candidates:
                return {"success": False, "message": "Failed to generate, Please try again! "}
            
            prompt_tokens = response.usage_metadata.prompt_token_count if hasattr(response, 'usage_metadata') else 0
            completion_tokens = response.usage_metadata.candidates_token_count if hasattr(response, 'usage_metadata') else 0
            total_tokens = prompt_tokens + completion_tokens if isinstance(prompt_tokens, int) else "N/A"
            
            print(f"LLM Tokens: Prompt={prompt_tokens or 'N/A'} | Result={completion_tokens or 'N/A'} | Total={total_tokens or 'N/A'}\n")

            result = response.text

            data = {
                'time': int(time.time()),
                'type': 'text',
                'count': 1,
                'tokens': total_tokens
            }
            self.api_key_records.append(data)
            self.redis_cli.set(self.request_key, json.dumps(self.api_key_records))
            
            return {
                "success": True, 
                "data": result, 
                "token_response": {
                    "prompt": prompt_tokens,
                    "generation": completion_tokens,
                    "total": total_tokens,
                },
                "message": "Generate done successfully!"
            }
        
        except Exception as e:
            print(f"\n Error: {str(e)}")
            return {"success": False, "message": str(e)}

    def upload_to_s3(self, image_data):
        try:
            
            s3_client = boto3.client(
                's3',
                aws_access_key_id=self.AWS_ACCESS_KEY,
                aws_secret_access_key=self.AWS_SECRET_ACCESS_KEY,
                region_name=self.AWS_REGION
            )

            image = Image.open(BytesIO((image_data)))
            # Convert PIL Image to bytes
            img_byte_arr = BytesIO()
            image.save(img_byte_arr, format='PNG')
            img_byte_arr = img_byte_arr.getvalue()

            # Generate unique filename
            
            filename = f"{uuid.uuid4()}.png"

            # Upload to S3
            s3_client.put_object(
                Bucket=self.AWS_BUCKET,
                Key=filename,
                Body=img_byte_arr,
                ContentType='image/png'
            )

            # Generate URL
            url = f"https://{self.AWS_BUCKET}.s3.{self.AWS_REGION}.amazonaws.com/{filename}"

            return url

        except Exception as e:
            print(f"Error uploading to S3: {str(e)}")
            return None
        
    def _generate_imge(self, request):
        
        try: 
            prompt = request.get('prompt', "")

            API_KEY = self.get_api_key()

            client = genai.Client(api_key=API_KEY)
            
            response = client.models.generate_content(
                model="gemini-2.0-flash-preview-image-generation",
                contents=prompt,
                config=types.GenerateContentConfig(
                    candidate_count=1,
                    response_modalities=['TEXT', 'IMAGE']
                ),
            )
            
            if not response.candidates:
                return {"success": False, "message": "Failed to generate, Please try again! "}
            
            prompt_tokens = response.usage_metadata.prompt_token_count if hasattr(response, 'usage_metadata') else 0
            completion_tokens = response.usage_metadata.candidates_token_count if hasattr(response, 'usage_metadata') else 0
            total_tokens = prompt_tokens + completion_tokens if isinstance(prompt_tokens, int) else "N/A"
            
            print(f"LLM Tokens: Prompt={prompt_tokens or 'N/A'} | Result={completion_tokens or 'N/A'} | Total={total_tokens or 'N/A'}\n")

            image_base_url = None

            for part in response.candidates[0].content.parts:
                if part.text is not None:
                    print(part.text)
                elif part.inline_data is not None:
                    image_base_url = self.upload_to_s3(part.inline_data.data)
            data = {
                'time': int(time.time()),
                'type': 'image',
                'count': 1,
                'tokens': total_tokens
            }
            self.api_key_records.append(data)
            self.redis_cli.set(self.request_key, json.dumps(self.api_key_records))

            return {
                "success": True, 
                "data": image_base_url, 
                "token_response": {
                    "prompt": prompt_tokens,
                    "generation": completion_tokens,
                    "total": total_tokens,
                },
                "message": "Generate done successfully!"
            }
        
        except Exception as e:
            print(f"\n Error: {str(e)}")
            return {"success": False, "message": str(e)}
    

File: backend/services/llm/gemini.js
const { GoogleGenerativeAI } = require("@google/generative-ai");
const AWS = require("aws-sdk");
const Base = require("./base");
const config = require("config");

const Utils = require("../../helpers/utils");

const NAME = "GeminiServices";
const MINUTE_SECONDS = 60;
const DAY_SECONDS = 86400;

class GeminiServices extends Base {
    constructor() {
        super(NAME);

        this.name = NAME;
        this.rpm = 10;
        this.rpd = 250;
        this.tpm = 250000;

        this.API_KEY = config.get("GOOGLE_GEMINI_APIKEY");
        this.API_KEYS = [config.get("GOOGLE_GEMINI_APIKEY")];
        this.MODEL = "gemini-2.5-flash";
        this.apiKeyRecords = [];

        this.AWS_ACCESS_KEY = config.get("AWS_ACCESS_KEY");
        this.AWS_SECRET_ACCESS_KEY = config.get("AWS_SECRET_ACCESS_KEY");
        this.AWS_REGION = config.get("AWS_REGION");
        this.AWS_BUCKET = config.get("AWS_BUCKET");

        // AWS S3 client
        this.s3Client = new AWS.S3({
            accessKeyId: this.AWS_ACCESS_KEY,
            secretAccessKey: this.AWS_SECRET_ACCESS_KEY,
            region: this.AWS_REGION,
        });
    }

    async getApiKey() {
        let apiKey = this.API_KEY;

        for (const key of this.API_KEYS) {
            const requestKey = `${this.name}_${key}_request_key`;
            const raw = await this.redisCli.get(requestKey);
            const data = JSON.parse(raw || "[]");

            this.apiKeyRecords = data;

            if (!data || data.length === 0) {
                apiKey = key;
            }

            let todayTotalRequests = 0;
            let currentMinRequests = 0;
            let currentMinTokens = 0;
            const now = Math.floor(Date.now() / 1000);

            for (const d of data) {
                if (d.time >= now - MINUTE_SECONDS) {
                    currentMinRequests += d.count;
                    currentMinTokens += d.tokens;
                }
                if (d.time >= now - DAY_SECONDS) {
                    todayTotalRequests += d.count;
                }
            }

            if (todayTotalRequests === 0 && currentMinRequests === 0 && currentMinTokens === 0) {
                apiKey = key;
            }

            if (todayTotalRequests > this.rpd || currentMinRequests > this.rpm || currentMinTokens > this.tpm) {
                continue;
            }

            apiKey = key;
        }

        return apiKey;
    }

    async _generate(request) {
        try {
            const prompt = request.prompt || "";
            const format = request.format || "text";

            let responseFormat = null;
            if (format === "json") {
                responseFormat = "application/json";
            } else if (format === "text") {
                responseFormat = "text/plain";
            }

            const API_KEY = await this.getApiKey();
            const client = new GoogleGenerativeAI(API_KEY);

            const model = client.getGenerativeModel({ model: this.MODEL });
            const result = await model.generateContent({
                contents: [{ role: "user", parts: [{ text: prompt }] }],
                generationConfig: { responseMimeType: responseFormat },
            });

            if (!result.response || !result.response.candidates?.length) {
                return { success: false, message: "Failed to generate, Please try again!" };
            }

            const usage = result.response.usageMetadata || {};
            const promptTokens = usage.promptTokenCount || 0;
            const completionTokens = usage.candidatesTokenCount || 0;
            const totalTokens = promptTokens + completionTokens;

            console.log(
                `LLM Tokens: Prompt=${promptTokens} | Result=${completionTokens} | Total=${totalTokens}`
            );

            const output = result.response.text();

            const data = {
                time: Math.floor(Date.now() / 1000),
                type: "text",
                count: 1,
                tokens: totalTokens,
            };

            this.apiKeyRecords.push(data);
            await this.redisCli.set(this.requestKey, JSON.stringify(this.apiKeyRecords));

            return {
                success: true,
                data: output,
                token_response: {
                    prompt: promptTokens,
                    generation: completionTokens,
                    total: totalTokens,
                },
                message: "Generate done successfully!",
            };
        } catch (e) {
            console.error("Error:", e.message);
            return { success: false, message: e.message };
        }
    }

    async uploadToS3(imageBuffer) {
        try {
            const filename = `${Utils.getUniqueId()}.png`;

            await this.s3Client
                .putObject({
                    Bucket: this.AWS_BUCKET,
                    Key: filename,
                    Body: imageBuffer,
                    ContentType: "image/png",
                })
                .promise();

            return `https://${this.AWS_BUCKET}.s3.${this.AWS_REGION}.amazonaws.com/${filename}`;
        } catch (e) {
            console.error("Error uploading to S3:", e.message);
            return null;
        }
    }

    async _generateImage(request) {
        try {
            const prompt = request.prompt || "";

            const API_KEY = await this.getApiKey();
            const client = new GoogleGenerativeAI(API_KEY);

            const model = client.getGenerativeModel({
                model: "gemini-2.0-flash-preview-image-generation",
            });

            const result = await model.generateContent({
                contents: [{ role: "user", parts: [{ text: prompt }] }],
                generationConfig: { candidateCount: 1 },
            });

            if (!result.response || !result.response.candidates?.length) {
                return { success: false, message: "Failed to generate, Please try again!" };
            }

            const usage = result.response.usageMetadata || {};
            const promptTokens = usage.promptTokenCount || 0;
            const completionTokens = usage.candidatesTokenCount || 0;
            const totalTokens = promptTokens + completionTokens;

            console.log(
                `LLM Tokens: Prompt=${promptTokens} | Result=${completionTokens} | Total=${totalTokens}`
            );

            let imageUrl = null;

            for (const part of result.response.candidates[0].content.parts) {
                if (part.text) {
                    console.log(part.text);
                } else if (part.inlineData) {
                    const buffer = Buffer.from(part.inlineData.data, "base64");
                    imageUrl = await this.uploadToS3(buffer);
                }
            }

            const data = {
                time: Math.floor(Date.now() / 1000),
                type: "image",
                count: 1,
                tokens: totalTokens,
            };

            this.apiKeyRecords.push(data);
            await this.redisCli.set(this.requestKey, JSON.stringify(this.apiKeyRecords));

            return {
                success: true,
                data: imageUrl,
                token_response: {
                    prompt: promptTokens,
                    generation: completionTokens,
                    total: totalTokens,
                },
                message: "Generate done successfully!",
            };
        } catch (e) {
            console.error("Error:", e.message);
            return { success: false, message: e.message };
        }
    }
}

module.exports = GeminiServices;
